### PR TITLE
feat: 🎸 add input masking component

### PR DESCRIPTION
**Pull request by codyarose**
Added a masking text field component for phone, zip, and currency

https://www.loom.com/share/637f18e7a37c45fe92a71c5fcbc49a08

### CODE REVIEW

**Review by lucas-homer — APPROVED** [review]
this all looks good to me. my only question is if we have any other common mask types we should support from the the outset. Deferring to @SeanGroff on that, though.

**Comment by 20BBrown14** [comment]
Seems like we could make this mask component general enough for the consumer to pass a custom mask? Or one of the default options provided.

**Comment by codyarose** [comment]
> Seems like we could make this mask component general enough for the consumer to pass a custom mask? Or one of the default options provided. 

Yeah, that's probably a good thing to add, otherwise the component user will have to request that a mask be added if it's not already an option

**Comment by SeanGroff** [comment]
Converting this PR to a Draft.

This is discussion-worthy but here's what I'm thinking initially. 

We leverage inversion of control. The MaskInput component accepts a custom mask only.

We then export on Object/Map containing pre-defined masks we can supply to the consumer. Then the consumer inserts the mask they import from SQForm into the MaskInput component.

Whenever we discover we need a new mask, we PR the new mask to or Masks Map. No changes to the MaskInput component would ever be required to allow for new masks.

**Comment by 20BBrown14** [comment]
I think like idea @SeanGroff personally. But one question. Are you suggesting that this component *should* only be used with the masks that we're exporting? And we won't necessarily support a custom mask being passed in. 

I was thinking that we can provide common masks (and add more if we decide to) but the consumer can pass custom masks with the help of the `text-masks` module instead of user one of ours out of the box.

Just trying to make sure I understand what you're envisioning.

**Comment by SeanGroff** [comment]
> I think like idea @SeanGroff personally. But one question. Are you suggesting that this component _should_ only be used with the masks that we're exporting? And we won't necessarily support a custom mask being passed in.
> 
> I was thinking that we can provide common masks (and add more if we decide to) but the consumer can pass custom masks with the help of the `text-masks` module instead of user one of ours out of the box.
> 
> Just trying to make sure I understand what you're envisioning.

So what I'm saying is the InputMaskComponent `mask` prop only accepts a "custom mask"  meaning it's up to the consumer to define the mask and pass it into the prop.

This custom prop can either be the (yet to be created) common masks we expose via the SQForm library, or a custom mask the developer comes up with for a special use case.

**Comment by 20BBrown14** [comment]
> This custom prop can either be the (yet to be created) common masks we expose via the SQForm library, or a custom mask the developer comes up with for a special use case.

Great. I love. This solution has my vote 👍

**Comment by codyarose** [comment]
Sounds good to me 👍
I suppose the existence of that masks object would be mentioned in the docs/story?

**Comment by SeanGroff** [comment]
> Sounds good to me 👍
> I suppose the existence of that masks object would be mentioned in the docs/story?

Precisely. In the future, we can add this to our Docs site for SQForm so it's more apparent.

File: src/components/SQForm/SQFormMaskedTextField.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import MaskedInput from 'react-text-mask';
+import SQFormTextField from './SQFormTextField';
+
+function TextFieldMask({inputRef, mask, ...rest}) {
+  return (
+    <MaskedInput
+      {...rest}
```
Comment (gdoohen1) [inline src/components/SQForm/SQFormMaskedTextField.js:9]: This looks awesome, I actually need something just like this =)
Could we maybe have `placeholderChar` and `showMask` be props with defaults that way they are able to be specified with inputProps and overridden if needed?

Comment (SeanGroff) [reply src/components/SQForm/SQFormMaskedTextField.js:9]: @gdoohen1 Could you clarify what `placeholderChar` and `showMask` would be used for.
Is this something that occurs in FlexForm you're needing? If so, can you screenshot or record a Loom to show the intent of these 2 props?

Comment (gdoohen1) [reply src/components/SQForm/SQFormMaskedTextField.js:9]: @SeanGroff I don't have a use-case at the moment to utilize these, I was more just asking from the perspective of erroring on the side of exposing something rather preventing the user from accessing it - but if we have strong UX standards of never showing masking and using the same placeholder than probably the extra level of safety in locking these up is the better option.

File: src/components/SQForm/SQFormMaskedTextField.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      inputProps={{
+        ...inputProps,
+        mask
+      }}
+      muiFieldProps={muiFieldProps}
+    />
+  );
+}
+
+SQFormMaskedTextField.propTypes = {
```
Comment (20BBrown14) [inline src/components/SQForm/SQFormMaskedTextField.js:63]: Our proptypes don't match the props we defined in the component definition.

File: src/utils/masks.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    /\d/,
+    /\d/,
+    /\d/,
+    '-',
+    /\d/,
+    /\d/,
+    /\d/,
+    /\d/
+  ],
+  zip: [/\d/, /\d/, /\d/, /\d/, /\d/],
```
Comment (20BBrown14) [inline src/utils/masks.js:20]: Would we also want a default for zip + 4?

File: src/components/SQForm/SQFormMaskedTextField.js (hunk 1 above)
Comment (SeanGroff) [reply src/components/SQForm/SQFormMaskedTextField.js:9]: Right on.
We try to keep the API small for ease of use then expand the props API as needs arise.
If the need arises for either of these features we'll enhance this component at that time.

File: src/utils/masks.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import createNumberMask from 'text-mask-addons/dist/createNumberMask';
+
+export const MASKS = {
+  phone: [
```
Comment (20BBrown14) [inline src/utils/masks.js:4]: Do we want a default phone mask that includes country code?

File: notes/SQForm.md [hunk 5]
```diff
@@ -56,6 +56,10 @@ As shown in the component prop-types, the `children` of this component must be a
 
 If you want an empty option in the dropdown list options, set the `displayEmpty` prop to `true`.
 
+**SQFormMaskedTextField**
```
Comment (20BBrown14) [inline notes/SQForm.md:59]: Could we perhaps mention the module we're using to help with the masks? A link would be great so someone using this component could do their own DD if needed.

File: src/components/SQForm/SQFormMaskedTextField.js (hunk 1 above)
Comment (gdoohen1) [reply src/components/SQForm/SQFormMaskedTextField.js:9]: Alright that sounds good to me!

File: src/components/SQForm/SQFormMaskedTextField.js (hunk 2 above)
Comment (codyarose) [reply src/components/SQForm/SQFormMaskedTextField.js:63]: Are you referring to the props that get passed to SQFormTextField?

Comment (20BBrown14) [reply src/components/SQForm/SQFormMaskedTextField.js:63]: Right

File: notes/SQForm.md (hunk 5 above)
Comment (codyarose) [reply notes/SQForm.md:59]: Good call, I'll add that

File: src/components/SQForm/SQFormMaskedTextField.js (hunk 2 above)
Comment (codyarose) [reply src/components/SQForm/SQFormMaskedTextField.js:63]: Gotcha, I'll add that. Still getting used to how propTypes is set up, I'm more familiar with typing props in TypeScript

**Comment by codyarose** [comment]
I'll defer to you all that are more familiar with how this would be used in production in regards to additional mask options, like the zip + 4 and phone country code @20BBrown14 mentioned. 

If you let me know what shape the mask should be in I'll work on adding it

**Comment by SeanGroff** [comment]
> I'll defer to you all that are more familiar with how this would be used in production in regards to additional mask options, like the zip + 4 and phone country code @20BBrown14 mentioned.
> 
> If you let me know what shape the mask should be in I'll work on adding it

We should steal the masks that already exist for flexform. I believe they live in a utils folder on the CRM's. Check sc2-senior.

**Review by 20BBrown14 — APPROVED** [review]

File: CHANGELOG.md [hunk 6]
```diff
@@ -7,342 +7,298 @@
 
 ### [4.1.1](https://github.com/SelectQuoteLabs/SQForm/compare/v4.1.0...v4.1.1) (2021-03-09)
 
-
 ### Bug Fixes
 
-* 🐛 make only dialog body scroll ([60beb7b](https://github.com/SelectQuoteLabs/SQForm/commit/60beb7b1e1bcf4da4fd01a2debb6fd50310504ae)), closes [#17](https://github.com/SelectQuoteLabs/SQForm/issues/17)
+- 🐛 make only dialog body scroll ([60beb7b](https://github.com/SelectQuoteLabs/SQForm/commit/60beb7b1e1bcf4da4fd01a2debb6fd50310504ae)), closes [#17](https://github.com/SelectQuoteLabs/SQForm/issues/17)
```
Comment (SeanGroff) [inline CHANGELOG.md:12]: This file is auto-generated. @codyarose can you dig into what might be causing this issue? My guess is Prettier for VS Code is formatting this markdown file..

Comment (SeanGroff) [reply CHANGELOG.md:12]: Spoke with Cody and he's going to dig into this

File: src/utils/masks.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
+import createNumberMask from 'text-mask-addons/dist/createNumberMask';
+import emailMask from 'text-mask-addons/dist/emailMask';
+
+export const MASKS = {
```
Comment (SeanGroff) [inline src/utils/masks.js:4]: Can we expose the existing REGEX masks we seem to copy/paste from project to project?
This way we ensure we expose masks that already meet the business's needs.
https://bitbucket.org/search?q=validationhelpers.js&account=%7Bf0e46ff7-117d-4015-9027-923794846585%7D

**Review by SeanGroff — CHANGES_REQUESTED** [review]
This is looking amazing. I'm excited to finally have masked fields in SQForm.
I requested to small changes. I'm good with merging this PR once the ChangeLog is removed from the git diff and the existing Regex masks we copy/pasta from project to project are exposed here.

File: src/utils/masks.js (hunk 7 above)
Comment (codyarose) [reply src/utils/masks.js:4]: @SeanGroff  Seems like those regex are formatted different from what `react-text-mask` expects. Should I gather the ones that are common across all those files and create an equivalent of them for `react-text-mask`, or are you suggesting we pull in all of that validation logic as well?

Comment (SeanGroff) [reply src/utils/masks.js:4]: > @SeanGroff Seems like those regex are formatted different from what `react-text-mask` expects. Should I gather the ones that are common across all those files and create an equivalent of them for `react-text-mask`, or are you suggesting we pull in all of that validation logic as well?

Gather the ones that are common across all those files and create an equivalent of them for `react-text-mask` would be ideal! I haven't looked at react-text-mask in a long time so I don't have a better suggestion.

I just want to ensure how we validate a phone or a zipcode remains the same as whats in production for several apps currently.

Comment (codyarose) [reply src/utils/masks.js:4]: Gotcha. I'll work on that

**Comment by codyarose** [comment]
@SeanGroff  Added an SSN mask, updated the zip mask to handle 5 or 5 + 4 digits, and added a little bit of regex validation to the phone number mask. Beyond that, the rest of the stuff in `validationHelpers` seems to be more for pure validation rather than masking.

I fear adding too much validation logic to the masks by preventing the user to input certain characters may end up making the inputs appear broken. Whereas being looser with what we allow the user to input and relying on yup validation for feedback we can let the user know that there's something wrong with what they've typed into the input rather than the input itself.

**Review by SeanGroff — APPROVED** [review]
Amazing @codyarose 🚀

**Pull request merged**